### ISSUE TITLE
Chains & Execution

### ISSUE THREAD

**Issue by splitice** [p0pr0ck5/FreeWAF#39]
It might be possible to get a greater execution speed if the control flow was pre-calculated.

What this would likely involve is precalculating a jump-offset for all rules, for non chained rules this would be one. In normal cases the pointer moves forward by one, either in a chain or when matching in a chain. If however the rule did not match the pointer moves forward by the rule's precalculated jump. That is it does to the end of the chain.

Saves a fair few executions of ```_process_rule```

**Comment by p0pr0ck5** [comment]
Yes, this is something I've thought about, it would be helpful if we decide to implement `header_filter_by_lua` phase handling to check upstream response data. Patches welcome :)

**Comment by splitice** [comment]
I also got to thinking that it might just be possible to represent chains the rules say like this:

```
{
{
var: ...
opt: ...
sub: {
var: 
opt:
...
}
},
{ next rule }
}
```

That should give the same effect, and is a bit simpler. Of course you can keep the same ruleset definitions and just build the nesting as a pre-process.

**Comment by p0pr0ck5** [comment]
I had been wanting to implement something like this, and it seems like it might be helpful for a bug brought up in issue #50. Patches welcome :)

**Comment by p0pr0ck5** [comment]
This is implemented in af63ef1. Performance gains from skipping over `_process_rule` are noticeable but not significant. We do gain the benefit of getting rid of some flags in the `ctx` table that previously managed logic for chaining and skipping.
